### PR TITLE
Effects

### LINKED ISSUE #1501: Effects like a Blur\Shadows
I want to make a card in Material Design that implies shadow effect. Are there any capabilities to do it?
image
WPF has a Effect Parameter, does Avalonia has something similar?

### PR DESCRIPTION
**Pull request by Sorien**
image

- What does the pull request do?

implements DropShadow effect (Skia, D2D), only for deferred rendering, i wanted to have it more like HTML than WPF so there are different properties

- How was the solution implemented (if it's not obvious)?

visual that contains effect will be rendered to separate layer, deferred renderer has one more layer (_effectLayer) that is used for applying effect (d2d cannot draw with a bitmap that is currently bound as the target bitmap) - don't like it and don't know how to solve it, this layer is cleared before applying clipping and effect otherwise there are visual artifacts

there are image page changes that should be reverted, also there are no unit tests, _i won't have a time to finish_ it so if anybody want to finish it you have my blessing

Checklist:

- [ ] Added unit tests (if possible)?
- [ ] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Avaloniaui.net with user documentation

Fixes #1501

### CODE REVIEW

**Comment by nc4rrillo** [comment]
Are you planning on adding offset, blur, color and spread properties to the drop shadow effect?

**Comment by Sorien** [comment]
i'll try to implement shadow and blur effect (skia and D2D) with all properties, but at first have to find right design, it's a bit hacky

**Comment by nc4rrillo** [comment]
Would it be possible to have this extend DrawingContext with something like `PushDropShadow(...)`/`PopDropShadow()`? that way you don't need the `IEffectImpl` and can just leverage the existing `DrawingContext` infrastructure

**Comment by Gillibald** [comment]
In my opinion an effect is just some manipulation to the rendered image. This should be implemented like a pipeline and transparent to all render calls. That way you can have several manipulations (Filter, Transform, etc) on a generated image. As far as I know Direct2D has a concept of composable effects. So you can have more than one effect like shadow blur etc. Each is an effect on its own.

**Comment by Gillibald** [comment]
I just thought a bit about this PR and checked how skia handles effects (filters). Filters are always applied to SkPaint in our case all drawings written to SkPaint (Canvas).

Can we add effects to our drawing context interface and use them for every drawing call?

To use multiple effects we can use a combined effect or a list of effects. I much prefer having a list and combine them if needed in the implementation.

**Comment by nc4rrillo** [comment]
Perhaps `PushOpacity` should become `PushFilters` with Opacity becoming one of many filters than can be pushed, others being DropShadow, Blur, Sepia, Grayscale, Invert, etc.

File: build/SharpDX.props [hunk 1]
```diff
@@ -4,6 +4,7 @@
     <PackageReference Include="SharpDX.Direct2D1" Version="4.0.1" />
     <PackageReference Include="SharpDX.Direct3D11" Version="4.0.1" />
     <PackageReference Include="SharpDX.DXGI" Version="4.0.1" />
+    <PackageReference Include="SharpDX.Mathematics" Version="4.0.1" />
```
Comment (jkoritzinsky) [inline build/SharpDX.props]: Make sure to add this package reference to the packages.cake build script.

**Comment by ryuujo1573** [comment]
Sorry I'm not good at English.. It might be rude but excuse me, I was wondering if this feature is implemented and if so how could I add it to a simple Border with adorable shadow effect?

**Comment by Gillibald** [comment]
BoxShadows are supported in the current master

**Comment by ryuujo1573** [comment]
Thank you so much!! It's so surprising to be replied so soon like this! THX!!

**Comment by graealex** [comment]
I'm a bid saddened that this effects pipeline never made it into a release. BoxShadow is nice, if you want a shadow. But not if you want to blur something.

**Pull request closed without merging**